cpp/src/parquet/arrow/reader_writer_benchmark.cc

```diff
@@ -280,6 +298,81 @@ BENCHMARK_TEMPLATE2(BM_ReadColumn, true, BooleanType)
     ->Args({kAlternatingOrNa, 1})
     ->Args({5, 10});
 
+//
+// Benchmark reading a nested column
+//
+
+static void BM_ReadStructColumn(::benchmark::State& state) {
+  constexpr int64_t kNumValues = BENCHMARK_SIZE / 10;
+  const double null_probability = static_cast<double>(state.range(0)) / 100.0;
+  const bool nullable = (null_probability != 0.0);
+
+  ARROW_CHECK_GE(null_probability, 0.0);
+
+  ::arrow::random::RandomArrayGenerator rng(42);
+
+  auto values1 = rng.Int32(kNumValues, -5, 5, null_probability);
+  auto values2 =
+      rng.Int64(kNumValues, -12345678912345LL, 12345678912345LL, null_probability);
+
+  const int64_t kBytesPerValue = sizeof(int32_t) + sizeof(int64_t);
+
+  std::shared_ptr<::arrow::Buffer> null_bitmap;
+  if (nullable) {
+    null_bitmap = rng.NullBitmap(kNumValues, null_probability);
+  }
+  auto array = *::arrow::StructArray::Make(
+      {values1, values2},
+      ::arrow::FieldVector{field("a", values1->type(), nullable),
+                           field("b", values2->type(), nullable)},
+      null_bitmap);
+  auto schema = ::arrow::schema({field("s", array->type(), nullable)});
+  auto table = ::arrow::Table::Make(schema, {array}, array->length());
+
+  EXIT_NOT_OK(table->Validate());
+
+  BenchmarkReadTable(state, *table, kNumValues, kBytesPerValue);
+}
+
+BENCHMARK(BM_ReadStructColumn)
+    ->Arg(/*null_percentage=*/0)
+    ->Arg(/*null_percentage=*/1)
+    ->Arg(/*null_percentage=*/50)
+    ->Arg(/*null_percentage=*/99);
+
+static void BM_ReadListColumn(::benchmark::State& state) {
+  constexpr int64_t kNumValues = BENCHMARK_SIZE / 10;
+  const double null_probability = static_cast<double>(state.range(0)) / 100.0;
+  const bool nullable = (null_probability != 0.0);
+
+  ARROW_CHECK_GE(null_probability, 0.0);
+
+  ::arrow::random::RandomArrayGenerator rng(42);
+
+  auto values = rng.Int64(kNumValues, -5, 5, null_probability);
+  auto offsets = rng.Offsets(kNumValues / 10, 0, values->length(), null_probability);
```

Comment:
approximately what size lists does this generate?

Reply:
`kNumValues / 10`

Reply:
so lists get larger as overall array goes up?

Reply:
Ah, you meant the size of individual lists? Well, since the list array will be `kNumValues / 10` large, individual lists should have 10 items average.

Reply:
ah, OK, I'm not sure if this is representative of real workloads but it is better then nothing.

Reply:
It's not really meant to be representative of anything real, just not outlandish :-)